### PR TITLE
Reset NetworkUnavailable condition on GCE

### PR DESCRIPTION
**Pull request by r0mant**
On GCE nodes come up marked with NetworkUnavailable condition which makes them unschedulable until this condition is reset. Normally, this condition gets reset by RouteController which we do not use. Flannel is essentially our "route controller" - when using GCE backend it creates routes itself. 

This PR makes flannel reset this node condition upon successful route creation. Looking at various issues on Github, that's what other people who make network plugins for Kubernetes do as well.

### CODE REVIEW

File: kube.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+// resetNodeCondition updates status of a Kubernetes node this flannel process
+// is running on to reset NetworkUnavailable condition
+//
+// On certain clouds, namely Google Compute Engine, nodes come up explicitly
+// marked with NetworkUnavailable condition which renders them unschedulable
+// and Kubernetes expects this condition to be removed either by RouteController
+// or networking plugins.
+//
+// We use neither of those: flannel does all route management (including
```
Comment (a-palchikov) [inline kube.go]: If this is ever meant to be merged upstream, there will be no "we" in the sense you mean. Maybe rephrase w/o this part:
```
// Since flannel is managing routes (including creating routes using cloud APIs using proper backend)
// it also needs to reset the condition.
```

File: main.go [hunk 2]
```diff
@@ -199,6 +209,14 @@ func main() {
 
 	// Start "Running" the backend network. This will block until the context is done so run in another goroutine.
 	go bn.Run(ctx)
+
+	// When running on GCE with cloud integration, the node comes up explicitly
+	// marked with NodeUnavailable status which flannel needs to reset after
+	// successfully provisioning a route on GCE
+	if config.BackendType == "gce" {
+		go resetNodeCondition(ctx)
```
Comment (a-palchikov) [inline main.go:217]: As far as I understood kubelet's code [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_node_status.go#L244), the node status is reset on each kubelet start which means, the `NodeNetworkUnavailable` will be reset if kubelet is restarted.
My question hence - do we need to patch flannel to do this or this can be off-site? We (could) have some reconiliation loops as part of the (planet) agent which would enforce this as I don't see anything in particular that ties this functionality to flannel.

File: kube.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+			if err != nil {
+				log.Warningf("Failed to create kube client: %v", err)
+				continue
+			}
+			patch, err := formatPatch()
+			if err != nil {
+				log.Warningf("Failed to format patch: %v", err)
+				continue
+			}
+			_, err = client.CoreV1().Nodes().PatchStatus(opts.kubeNode, patch)
```
Comment (a-palchikov) [inline kube.go:50]: You'd only need to do this upon encountering `NodeNetworkUnavailable` (and not unconditionally) and hence only when the kubelet has actually started (since it's the one setting the status) - otherwise you run the risk of patching it before kubelet resets it.

File: main.go (hunk 2 above)
Comment (r0mant) [reply main.go:217]: No, this status is set only during initial node registration. It is not reset when kubelet restarts.

And it actually ties to flannel directly because it's flannel who creates routes. If it fails to provision a route, the node will keep NetworkUnavailable status and rightfully so.

File: kube.go (hunk 3 above)
Comment (r0mant) [reply kube.go:50]: As I mentioned above, this status is only set during initial node registration. When kubelet starts for the first time, the node is not registered yet and hence this code will be spinning. Then kubelet will register the node which already includes NetworkUnavailable status, so it will be on the node right from the start.

Basically, what I'm trying to say is the race you're describing can't happen, so I don't think it makes sense to overcomplicate with trying to check the status first. If flannel restarts, it will update it again, yes, but that's fine - it's kind of like an "upsert".

**Review by a-palchikov — APPROVED** [review]
